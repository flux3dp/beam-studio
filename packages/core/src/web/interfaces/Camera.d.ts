export interface CameraParameters {
  angle: number;
  scaleRatioX: number;
  scaleRatioY: number;
  x: number;
  y: number;
}

export interface CameraConfig {
  R: number;
  SX: number;
  SY: number;
  X: number;
  Y: number;
}
