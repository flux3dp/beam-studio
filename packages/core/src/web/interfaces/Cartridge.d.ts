export interface ChipSettings {
  brand: string;
  color: number;
  offset: number[];
  plScale: number;
  serial: string;
  timeUsed: number;
  totalCapacity: number;
  type: number;
  uid: string;
  verified?: boolean;
}

export interface RawChipSettings {
  brand?: string;
  color?: number;
  pl_scale?: number;
  pos_offset?: number[];
  serial?: string;
  total_capacity?: number;
  type?: number;
  uid?: string;
  used_time?: number;
  verified?: boolean;
}
