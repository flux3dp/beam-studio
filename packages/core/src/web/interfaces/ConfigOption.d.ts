interface ConfigOption {
  label?: string;
  value: number;
}

export default ConfigOption;
