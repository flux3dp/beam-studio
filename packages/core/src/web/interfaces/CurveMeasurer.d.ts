import type { MeasureData } from './ICurveEngraving';

export interface InteractiveOptions {
  checkCancel?: () => boolean;
  onPointFinished?: (finishedCount: number) => void;
  onProgressText?: (text: string) => void;
}

export type MeasurePointData = { height: null | number; xOffset?: number; yOffset?: number };

export interface CurveMeasurer {
  end: () => Promise<void>;
  measureArea: (
    xRange: number[],
    yRange: number[],
    objectHeight: number,
    opts?: InteractiveOptions,
  ) => Promise<MeasureData | null>;
  measurePoint: (
    x: number,
    y: number,
    feedrate: number,
    offset?: [number, number, number],
    objectHeight?: number,
    lowest?: number,
  ) => Promise<MeasurePointData>;
  measurePoints: (
    curData: MeasureData,
    targetIndices: number[],
    opts?: InteractiveOptions,
  ) => Promise<MeasureData | null>;
  setup: (onProgressText?: (text: string) => void) => Promise<boolean>;
  setupDevice: () => Promise<boolean>;
}
