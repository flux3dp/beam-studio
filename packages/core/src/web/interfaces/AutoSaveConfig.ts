export interface AutoSaveConfig {
  directory: string;
  enabled: boolean;
  fileNumber: number;
  timeInterval: number;
}
