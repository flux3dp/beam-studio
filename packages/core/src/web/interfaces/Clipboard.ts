export interface ClipboardElement {
  attributes: Array<Record<'namespaceURI' | 'nodeName' | 'value', null | string>>;
  childNodes: ClipboardElement[];
  dataGSVG?: string;
  dataSymbol?: string;
  innerHTML: string;
  namespaceURI: null | string;
  nodeName: string;
  nodeType: number;
  nodeValue: null | string;
}

export interface ClipboardData {
  elements: ClipboardElement[];
  imageData: Record<string, string>;
  refs: Record<string, ClipboardElement>;
  // The source of the clipboard data, e.g., the current tab ID.
  // This can be used to identify where the clipboard data originated from.
  source: string;
}

export interface ClipboardCore {
  addRefToClipboard(useElement: SVGUseElement): void;

  /**
   * Copy the given elements to the clipboard.
   * @param elems The elements to copy.
   */
  copyElements(elems: Element[]): Promise<void>;

  /**
   * Get the data from the clipboard.
   */
  getData(): Promise<Element[]>;

  /**
   * Get the raw data from the clipboard.
   */
  getRawData(): Promise<ClipboardData | null>;

  getRefFromClipboard(id: string): Element | undefined;

  hasData(): Promise<boolean>;
}
