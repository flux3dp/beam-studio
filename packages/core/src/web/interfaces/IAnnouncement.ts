export interface IAnnouncement {
  content: string;
  id: number;
  link: string;
  link_text: string;
  title: string;
}
