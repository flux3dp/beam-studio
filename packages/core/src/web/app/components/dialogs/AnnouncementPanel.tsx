import React, { useState } from 'react';
import { Button, Checkbox, Modal } from 'antd';

import AnnouncementHelper from 'helpers/announcement-helper';
import browser from 'implementations/browser';
import useI18n from 'helpers/useI18n';
import { ExportOutlined } from '@ant-design/icons';
import { IAnnouncement } from 'interfaces/IAnnouncement';

import styles from './AnnouncementPanel.module.scss';

interface Props {
  announcement: IAnnouncement;
  onClose: () => void;
}

const AnnouncementPanel = ({ announcement, onClose }: Props): JSX.Element => {
  const lang = useI18n();
  const [isCheckboxChecked, setIsCheckboxChecked] = useState(false);

  const handleClose = () => {
    if (isCheckboxChecked) AnnouncementHelper.setNotShowing(announcement.id);
    onClose();
  };

  return (
    <Modal
      className={styles.announcement}
      title={announcement.title || lang.beambox.announcement_panel.title}
      onCancel={handleClose}
      onOk={handleClose}
      footer={
        announcement.link ? (
          <Button type="primary" onClick={() => browser.open(announcement.link)}>
            {announcement.link_text || lang.alert.learn_more} <ExportOutlined />
          </Button>
        ) : (
          <Button type="primary" onClick={handleClose}>
            {lang.alert.close}
          </Button>
        )
      }
      centered
      open
    >
      <div className="main-content">
        {/* eslint-disable-next-line react/no-danger */}
        <div dangerouslySetInnerHTML={{ __html: announcement.content }} />
        <div className={styles.checkbox}>
          <Checkbox
            checked={isCheckboxChecked}
            onChange={() => setIsCheckboxChecked(!isCheckboxChecked)}
          >
            {lang.beambox.announcement_panel.dont_show_again}
          </Checkbox>
        </div>
      </div>
    </Modal>
  );
};

export default AnnouncementPanel;
