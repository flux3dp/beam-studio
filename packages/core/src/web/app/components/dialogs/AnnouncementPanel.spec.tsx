import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import { IAnnouncement } from 'interfaces/IAnnouncement';

import AnnouncementPanel from './AnnouncementPanel';

const mockAnnouncement: IAnnouncement = {
  id: 11,
  title: 'mock title',
  content: 'mock content',
  link: 'https://example.com',
  link_text: 'mock link content',
};

jest.mock('helpers/useI18n', () => () => ({
  alert: {
    learn_more: 'Learn More',
    close: 'Close',
  },
  beambox: {
    announcement_panel: {
      title: 'Announcement',
      dont_show_again: "Don't show again",
    },
  },
}));

const mockOpen = jest.fn();
jest.mock('implementations/browser', () => ({
  open: (...args) => mockOpen(...args),
}));

const mockSetNotShowing = jest.fn();
jest.mock('helpers/announcement-helper', () => ({
  setNotShowing: (...args) => mockSetNotShowing(...args),
}));

const mockOnClose = jest.fn();

describe('test AnnouncementPanel', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { baseElement, getByText } = render(
      <AnnouncementPanel announcement={mockAnnouncement} onClose={mockOnClose} />
    );
    expect(baseElement).toMatchSnapshot();

    fireEvent.click(getByText('mock link content'));
    expect(mockOpen).toHaveBeenCalledTimes(1);
    expect(mockOpen).toHaveBeenLastCalledWith('https://example.com');

    fireEvent.click(baseElement.querySelector('.ant-modal-close'));
    expect(mockOnClose).toHaveBeenCalledTimes(1);
    expect(mockSetNotShowing).not.toHaveBeenCalled();

    const checkbox = baseElement.querySelector('.ant-checkbox');
    expect(checkbox).not.toHaveClass('ant-checkbox-checked');
    fireEvent.click(getByText("Don't show again"));
    expect(checkbox).toHaveClass('ant-checkbox-checked');
    fireEvent.click(baseElement.querySelector('.ant-modal-close'));
    expect(mockOnClose).toHaveBeenCalledTimes(2);
    expect(mockSetNotShowing).toHaveBeenCalledTimes(1);
    expect(mockSetNotShowing).toHaveBeenCalledWith(11);
  });

  it('should render correctly without button link', () => {
    const { baseElement, getByText } = render(
      <AnnouncementPanel announcement={{ ...mockAnnouncement, link: '' }} onClose={mockOnClose} />
    );
    expect(baseElement).toMatchSnapshot();

    const closeButton = getByText('Close');
    fireEvent.click(closeButton);
    expect(mockOpen).not.toHaveBeenCalled();
    expect(mockOnClose).toHaveBeenCalledTimes(1);
    expect(mockSetNotShowing).not.toHaveBeenCalled();

    fireEvent.click(getByText("Don't show again"));
    expect(mockSetNotShowing).toHaveBeenCalledTimes(0);
    fireEvent.click(closeButton);
    expect(mockOnClose).toHaveBeenCalledTimes(2);
    expect(mockSetNotShowing).toHaveBeenCalledTimes(1);
    expect(mockSetNotShowing).toHaveBeenCalledWith(11);
  });
});
