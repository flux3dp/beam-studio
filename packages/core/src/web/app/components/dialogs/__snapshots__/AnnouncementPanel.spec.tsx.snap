// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test AnnouncementPanel should render correctly 1`] = `
<body>
  <div />
  <div>
    <div
      class="ant-modal-root css-dev-only-do-not-override-hash"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        class="ant-modal-wrap ant-modal-centered"
        tabindex="-1"
      >
        <div
          aria-labelledby="test-id"
          aria-modal="true"
          class="ant-modal css-dev-only-do-not-override-hash announcement ant-zoom-appear ant-zoom-appear-prepare ant-zoom"
          role="dialog"
          style="width: 520px;"
        >
          <div
            style="outline: none;"
            tabindex="0"
          >
            <div
              class="ant-modal-content"
            >
              <button
                aria-label="Close"
                class="ant-modal-close"
                type="button"
              >
                <span
                  class="ant-modal-close-x"
                >
                  <span
                    aria-label="close"
                    class="anticon anticon-close ant-modal-close-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close"
                      fill="currentColor"
                      fill-rule="evenodd"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
              <div
                class="ant-modal-header"
              >
                <div
                  class="ant-modal-title"
                  id="test-id"
                >
                  mock title
                </div>
              </div>
              <div
                class="ant-modal-body"
              >
                <div
                  class="main-content"
                >
                  <div>
                    mock content
                  </div>
                  <div
                    class="checkbox"
                  >
                    <label
                      class="ant-checkbox-wrapper css-dev-only-do-not-override-hash"
                    >
                      <span
                        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-hash"
                      >
                        <input
                          class="ant-checkbox-input"
                          type="checkbox"
                        />
                        <span
                          class="ant-checkbox-inner"
                        />
                      </span>
                      <span>
                        Don't show again
                      </span>
                    </label>
                  </div>
                </div>
              </div>
              <div
                class="ant-modal-footer"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-hash ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                  type="button"
                >
                  <span>
                    mock link content 
                  </span>
                  <span
                    aria-label="export"
                    class="anticon anticon-export"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="export"
                      fill="currentColor"
                      fill-rule="evenodd"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M880 912H144c-17.7 0-32-14.3-32-32V144c0-17.7 14.3-32 32-32h360c4.4 0 8 3.6 8 8v56c0 4.4-3.6 8-8 8H184v656h656V520c0-4.4 3.6-8 8-8h56c4.4 0 8 3.6 8 8v360c0 17.7-14.3 32-32 32zM770.87 199.13l-52.2-52.2a8.01 8.01 0 014.7-13.6l179.4-21c5.1-.6 9.5 3.7 8.9 8.9l-21 179.4c-.8 6.6-8.9 9.4-13.6 4.7l-52.4-52.4-256.2 256.2a8.03 8.03 0 01-11.3 0l-42.4-42.4a8.03 8.03 0 010-11.3l256.1-256.3z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
          </div>
          <div
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`test AnnouncementPanel should render correctly without button link 1`] = `
<body>
  <div />
  <div>
    <div
      class="ant-modal-root css-dev-only-do-not-override-hash"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        class="ant-modal-wrap ant-modal-centered"
        tabindex="-1"
      >
        <div
          aria-labelledby="test-id"
          aria-modal="true"
          class="ant-modal css-dev-only-do-not-override-hash announcement ant-zoom-appear ant-zoom-appear-prepare ant-zoom"
          role="dialog"
          style="width: 520px;"
        >
          <div
            style="outline: none;"
            tabindex="0"
          >
            <div
              class="ant-modal-content"
            >
              <button
                aria-label="Close"
                class="ant-modal-close"
                type="button"
              >
                <span
                  class="ant-modal-close-x"
                >
                  <span
                    aria-label="close"
                    class="anticon anticon-close ant-modal-close-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close"
                      fill="currentColor"
                      fill-rule="evenodd"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
              <div
                class="ant-modal-header"
              >
                <div
                  class="ant-modal-title"
                  id="test-id"
                >
                  mock title
                </div>
              </div>
              <div
                class="ant-modal-body"
              >
                <div
                  class="main-content"
                >
                  <div>
                    mock content
                  </div>
                  <div
                    class="checkbox"
                  >
                    <label
                      class="ant-checkbox-wrapper css-dev-only-do-not-override-hash"
                    >
                      <span
                        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-hash"
                      >
                        <input
                          class="ant-checkbox-input"
                          type="checkbox"
                        />
                        <span
                          class="ant-checkbox-inner"
                        />
                      </span>
                      <span>
                        Don't show again
                      </span>
                    </label>
                  </div>
                </div>
              </div>
              <div
                class="ant-modal-footer"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-hash ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                  type="button"
                >
                  <span>
                    Close
                  </span>
                </button>
              </div>
            </div>
          </div>
          <div
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
